### PR TITLE
Introduce cache to hold uncommitted log entries.

### PR DESCRIPTION
**Pull request by maxsumrall**
Log entries which are uncommitted and waiting for cluster members to
 receive the entry are stashed in the cache for the commit step.

### CODE REVIEW

**Comment by maxsumrall** [comment]
Interestingly, it passes locally when there are no elections. In this failed build, there was a election and the first entry it cannot find is a NewLeaderBarrier.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/InFlightMap.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * Adds a new mapping.
+     *
+     * @param key The key of the mapping
+     * @param value The value corresponding to the key provided.
+     * @throws IllegalArgumentException if a mapping for the key already exists
+     */
+    public void register( K key, V value )
+    {
+        V previousValue = map.put( key, value );
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/InFlightMap.java]: `putIfAbsent`?

**Comment by davidegrohmann** [comment]
@maxsumrall are you working on this?

**Comment by maxsumrall** [comment]
It is on the stack of things to do.

**Comment by maxsumrall** [comment]
Will rebase and review after https://github.com/neo4j/neo4j/pull/7262 is merged which fixes the issue originally found.

**Comment by davidegrohmann** [comment]
@maxsumrall needs rebase

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 2]
```diff
@@ -117,24 +114,42 @@ public synchronized void notifyCommitted( long commitIndex )
...
-                        maybeFlush();
+                    if ( raftLogEntry == null )
+                    {
+                        String failureMessage = format(
+                                "Log entry at index %d is not present in the log. Please perform recovery by " +
+                                        "restarting this database instance. Indexes to be applied are %d up to %d, " +
+                                        "append index is %d, prev index is %d.",
+                                i, lastApplied + 1, lastToApply, raftLog.appendIndex(), raftLog.prevIndex() );
+                        log.error( failureMessage );
+                        throw new IllegalStateException( failureMessage );
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: let's clean up this code it is really hard to read

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/InFlightMap.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import static java.lang.String.format;
+
+public class InFlightMap<K, V>
+{
+    private final Map<K,V> map;
+
+    public InFlightMap()
+    {
+        this.map = new ConcurrentHashMap<>();
+    }
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/InFlightMap.java]: let's delete the constructor and initialize the field inline

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 2 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: I think having a `DelegateRaftLog` class that will try to use the cache first or call the real raftlog implementation is a better way to abstract the cache away while reading

**Pull request merged**